docs/ffi/ffi.md

```diff
@@ -0,0 +1,467 @@
+So, here's a simple implementation of RMS normalization in C++:
+
+```c++
+#include <cmath>
+#include <cstdint>
+
+float ComputeRmsNorm(float eps, int64_t size, const float *x, float *y) {
+  float sm = 0.0f;
+  for (int64_t n = 0; n < size; ++n) {
+    sm += x[n] * x[n];
+  }
+  float scale = 1.0f / std::sqrt(sm / float(size) + eps);
+  for (int64_t n = 0; n < size; ++n) {
+    y[n] = x[n] * scale;
+  }
+  return scale;
+}
+```
+
+and, for our example, this is the function that we want to expose to JAX via the FFI.
+
+### C++ interface
+
+To expose our library function to JAX and XLA, we need to write a thin wrapper using the APIs provided by the header-only library in the [`xla/ffi/api`](https://github.com/openxla/xla/tree/main/xla/ffi/api) directory of the [XLA project](https://github.com/openxla/xla).
+For more information about this interface, take a look at [the XLA custom call documentation](https://openxla.org/xla/custom_call).
+The full source listing can be downloaded [here](rms_norm.cc), but the key implementation details are reproduced here:
+
+```c++
+#include <functional>
+#include <numeric>
+#include <utility>
+
+#include "xla/ffi/api/c_api.h"
+#include "xla/ffi/api/ffi.h"
+
+namespace ffi = xla::ffi;
+
+std::pair<int64_t, int64_t> GetDims(ffi::Span<const int64_t> dims) {
+  if (dims.size() == 0) {
+    return std::make_pair(0, 0);
+  }
+  int64_t totalSize =
+    std::accumulate(dims.begin(), dims.end(), 1, std::multiplies<>());
+  int64_t lastDim = dims.back();
+  return std::make_pair(totalSize, lastDim);
+}
+
+ffi::Error RmsNormImpl(float eps, ffi::Buffer<ffi::DataType::F32> x,
+                       ffi::Result<ffi::Buffer<ffi::DataType::F32>> y) {
+  auto [totalSize, lastDim] = GetDims(x.dimensions);
+  if (lastDim == 0) {
+    return ffi::Error(ffi::ErrorCode::kInvalidArgument,
+                      "RmsNorm input must be an array");
+  }
+  for (int64_t n = 0; n < totalSize; n += lastDim) {
+    ComputeRmsNorm(eps, lastDim, &(x.data[n]), &(y->data[n]));
+  }
+  return ffi::Error::Success();
+}
+
+XLA_FFI_DEFINE_HANDLER(
+    RmsNorm, RmsNormImpl,
+    ffi::Ffi::Bind()
+        .Attr<float>("eps")
+        .Arg<ffi::Buffer<ffi::DataType::F32>>(/* x */)
+        .Ret<ffi::Buffer<ffi::DataType::F32>>(/* y */));
+```
+
+Starting at the bottom, we're using the XLA-provided macro `XLA_FFI_DEFINE_HANDLER` to generate some boilerplate which will expand into a function called `RmsNorm` with the appropriate signature.
+But, the important stuff here is all in the call to `ffi::Ffi::Bind()`, where we define the input and output types, and the types of any parameters.
+
+Then, in `RmsNormImpl`, we accept `ffi::Buffer` arguments which include information about the buffer shape, and pointers to the underlying data.
+In this implementation, we treat all leading dimensions of the buffer as batch dimensions, and perform RMS normalization over the last axis.
```

Comment:
I find "all leading dimensions" to be ambiguous. For some kernels all but the last 2 dimension should be the batch dimensions, and for others perhaps all but the last one.

Reply:
I agree! I've added a bit more discussion and a link to the `vmap` section below hoping that this can help clarify things.